### PR TITLE
#109: Gallery deletion crash

### PR DESCRIPTION
**Pull request by CloudLevi**
Needs further testing. Removed unnecessary loading, and the crash when deleting a resource and going back is fixed.

### CODE REVIEW

File: app/src/main/java/space/taran/arknavigator/mvp/presenter/ResourcesPresenter.kt [hunk 1]
```diff
@@ -116,23 +116,23 @@ class ResourcesPresenter(
 
     fun onViewResume() {
         tagsSelectorPresenter.calculateTagsAndSelection()
-        if (!tagsEnabled) {
-            presenterScope.launch {
-                resetResources(listResources(untaggedOnly = true))
-            }
+        if (this::storage.isInitialized && this::index.isInitialized) {
+            val aggregatedIndex = (index as? AggregatedResourcesIndex)?: return
```
Comment (kirillt) [inline app/src/main/java/space/taran/arknavigator/mvp/presenter/ResourcesPresenter.kt:120]: 1) This is rude break of incapsulation principles.
2) This is not robust logic: when you have an interface you must assume the object to be any of possible implementations. Here, the bug is fixed for sharded index, but not for plain one.
3) Don't touch indexes please, this removal logic doesn't belong to them.

**Review by kirillt — CHANGES_REQUESTED** [review]
Looks like dirty hack to me. Here, you should establish communication between UI elements, without involving lower-level components. Casting of object given by interface to particular implementation is really bad practice (read McConnell please, or some newer book about design and programming patterns, also about incapsulation). Even if we wanted to use index to notify all dependent objects about removals, that must be implemented more carefully (with storing removed ids in shards and aggregating them in sharded index), but that's redundant anyway.

**Review by mdrlzy — COMMENTED** [review]
The main reason for the crash is asynchronous update of the selection. That's why RecyclerView binds elements from the old list.

As it seems to me, there are 2 solutions here:
- `index.getPath` should return a nullable result
- wrap `index.getPath` in try-catch

**Comment by kirillt** [comment]
Might be possible to just wait till updating coroutine finished and bind elements with RecyclerView only afterwards.

File: app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/AggregatedResourcesIndex.kt [hunk 2]
```diff
@@ -7,6 +7,8 @@ class AggregatedResourcesIndex(
     private val shards: Collection<PlainResourcesIndex>)
```
Comment (darkovski88) [inline app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/AggregatedResourcesIndex.kt:7]: maybe add some kind of documentation so when other people read the code can know what this class is actually for?

File: app/src/main/java/space/taran/arknavigator/mvp/presenter/ResourcesPresenter.kt (hunk 1 above)
Comment (darkovski88) [reply app/src/main/java/space/taran/arknavigator/mvp/presenter/ResourcesPresenter.kt:120]: I have one thing to add here - why do we have a presenter within a presenter? What is the reason behind this?

Comment (kirillt) [reply app/src/main/java/space/taran/arknavigator/mvp/presenter/ResourcesPresenter.kt:120]: @darkovski88 good question, maybe @mdrlzy can answer

File: app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/AggregatedResourcesIndex.kt (hunk 2 above)
Comment (kirillt) [reply app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/AggregatedResourcesIndex.kt:7]: @darkovski88 yeah, we need to document core classes.
Maybe this will be what we add:
```
`AggregatedResourceIndex` is useful for "aggregated" navigation mode — a mode in which we navigate
through multiple indexed folders (roots). For a single-root navigation `PlainResourceIndex` can be used, and
both `AggregatedResourceIndex` and `PlainResourceIndex` should be transparently interchangeable,
meaning that both implementations support the same methods and they should not be used
otherwise as by `ResourceIndex` interface. Any component using `ResourceIndex` should
function in the same fashion independent on implementation of the index,
and only capability to look into multiple roots is gained
by passing `AggregatedResourceIndex` into the component. 
```
I am not sure if it's too wordy or not, I hope it explains the purpose of this class.

File: app/src/main/java/space/taran/arknavigator/mvp/presenter/ResourcesPresenter.kt (hunk 1 above)
Comment (mdrlzy) [reply app/src/main/java/space/taran/arknavigator/mvp/presenter/ResourcesPresenter.kt:120]: Multiple presenters are needed to separate business logic.
`gridPresenter` is responsible for which resources will be displayed, also handling the click on the resource and sorting.
`tagsSelectorPresenter` is responsible for which tags will be displayed, also handling a click on a tag and calculating a selection.
`ResourcesPresenter` is a screen layer presenter, it is responsible for initializing the index, tags storage, and handling toolbar menu click. It is also needed for communication between `gridPresenter` and `tagsSelectorPresenter`.

**Comment by kirillt** [comment]
Substituted with https://github.com/ARK-Builders/ARK-Navigator/pull/172
